schema_enforcer/instances/file.py

```diff
@@ -62,53 +71,97 @@ def __init__(self, root, filename, matches=None):
         Args:
             root (string): Absolute path to the directory where the schema file is located.
             filename (string): Name of the file.
-            matches (list, optional): List of schema IDs that matches with this Instance file. Defaults to None.
+            matches (set, optional): Set of schema IDs that matches with this Instance file. Defaults to None.
         """
         self.data = None
         self.path = root
         self.full_path = os.path.realpath(root)
         self.filename = filename
 
+        # Internal vars for caching data
+        self._top_level_properties = set()
+
         if matches:
             self.matches = matches
         else:
-            self.matches = []
+            self.matches = set()
+
+        self._add_matches_by_decorator()
+
+    @property
+    def top_level_properties(self):
+        """Return a list of top level properties in the structured data defined by the data pulled from _get_content.
+
+        Returns:
+            top_level_properties (set): Set of the strings of top level properties defined by the data file
+        """
+        if self._top_level_properties:
+            return self._top_level_properties
 
-        self.matches.extend(self._find_matches_inline())
+        content = self._get_content()
+        top_level_properties = set(content.keys())
+        return top_level_properties
```

Comment:
Is it intentional that if `self._top_level_properties` is an empty set, it never updates it with `self._top_level_properties = top_level_properties`?

More generally, for properties, I like/recommend the following single-return approach:

```python
@property
def my_property(self):
    if not self._my_property:
        # do whatever needs to be done to initialize the property...
        self._my_property = some_value
    return self._my_property
```